openwisp_monitoring/device/apps.py

```diff
@@ -73,3 +77,63 @@ def trigger_device_recovery_checks(cls, instance, **kwargs):
 
         if cache.get(get_device_recovery_cache_key(device=instance)):
             trigger_device_recovery.delay(pk=instance.pk)
+
+    @classmethod
+    def connect_is_working_changed(self):
+        from openwisp_controller.connection.models import DeviceConnection
+        from openwisp_controller.connection.signals import is_working_changed
+
+        is_working_changed.connect(
+            self.is_working_changed_receiver,
+            sender=DeviceConnection,
+            dispatch_uid='change_health_status',
+        )
+
+    @classmethod
+    def is_working_changed_receiver(cls, instance, is_working, **kwargs):
+        from ..check.models import Check
+        from .tasks import trigger_device_recovery
+
+        device = instance.device
+        device_monitoring = device.monitoring
+        initial_status = device_monitoring.status
+        status = 'ok' if is_working else 'problem'
+        if not is_working:
+            device_monitoring.update_status(status)
```

Comment:
The code in `trigger_device_recovery` also changes the status so we may be changing the status twice in a matter of milliseconds. 
I would delegate the health status change entirely to that task, it doesn't matter if we flag it as problem or critical, we'll fine tune this in the future as everything becomes clear.

I would rename the task to `trigger_device_checks` and just call that, eg:

```python
trigger_device_recovery.delay(pk=device.pk)
# ... prepare notification data ... 
if not is_working:
   # ... prepare notification data
else:
   # ...
notify_users()
```

Reply:
@nemesisdesign, in the issue description it was mentioned to trigger checks only on specific instances like:
If device is not working & previous status was working only then trigger checks
While in your suggestion above related `checks` are triggered for all cases so what is required?